File: Smaller Sum[ 22-April-2023]/HashMap based Approach/Solution.java
//   >> JAVA CODE <<


public long[] smallerSum(int n,int arr[])
    {
        HashMap<Integer,Long> map=new HashMap<>();
        long[] ans=new long[n]; int in=0;
        int[] arrcpy=new int[n];
        
        for(int i=0;i<n;i++) arrcpy[i]=arr[i]; 
        
        
        Arrays.sort(arrcpy);
        
        long prefixSum=arrcpy[0];
        map.put(arrcpy[0],(long)0);
        
        for(int i=1;i<n;i++)
        {
            if(arrcpy[i-1]==arrcpy[i])
            {
                long sum=map.get(arrcpy[i-1]);
                map.put(arrcpy[i],sum);
            }
            else
            {
                map.put(arrcpy[i],prefixSum);
            }
            
            prefixSum+=arrcpy[i];
        }
        
        
        for(int i=0;i<n;i++)
        {
            ans[i]=map.get(arr[i]);
        }
        
            return ans;
    }

File: Smaller Sum[ 22-April-2023]/HashMap based Approach/solution.py
#   >> PYTHON CODE <<


class Solution:
    
    def smallerSum(self, n : int, arr : List[int]) -> List[int]:
        hashmap = {}
        ans = []
        arrcpy = arr.copy()
        
        arrcpy.sort()
        
        prefixSum = arrcpy[0]
        hashmap[arrcpy[0]] = 0
        
        for i in range(1, n):
            if arrcpy[i-1] == arrcpy[i]:
                sum = hashmap[arrcpy[i-1]]
                hashmap[arrcpy[i]] = sum
            else:
                hashmap[arrcpy[i]] = prefixSum
            
            prefixSum += arrcpy[i]
        
        for i in range(n):
            ans.append(hashmap[arr[i]])
        
        return ans